src/System.Threading/tests/InterlockedTests.netcoreapp.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Threading.Tests
+{
+    public partial class InterlockedTests
+    {
+        // Taking this lock on the same thread repeatedly is very fast because it has no interlocked operations. 
+        // Switching the thread where the lock is taken is expensive because of allocation and FlushProcessWriteBuffers.
+        private class AsymmetricLock
+        {
+            public class LockCookie
+            {
+                internal LockCookie(int threadId)
+                {
+                    ThreadId = threadId;
+                    Taken = false;
+                }
+
+                public void Exit()
+                {
+                    Debug.Assert(ThreadId == Environment.CurrentManagedThreadId);
+                    Taken = false;
+                }
+
+                internal readonly int ThreadId;
+                internal bool Taken;
+            }
+
+            private LockCookie _current = new LockCookie(-1);
+
+            // Returning LockCookie to call Exit on is the fastest implementation because of it works naturally with the RCU pattern.
+            // The traditional Enter/Exit lock interface would require thread local storage or some other scheme to reclaim the cookie.
+            public LockCookie Enter()
+            {
+                int currentThreadId = Environment.CurrentManagedThreadId;
+
+                LockCookie entry = _current;
+
+                if (entry.ThreadId == currentThreadId)
+                {
+                    entry.Taken = true;
+
+                    //
+                    // If other thread started stealing the ownership, we need to take slow path.
+                    //
+                    // Volatile works here, but it is too big of a hammer because of it will result into memory barrier on ARM that 
+                    // we do not need here. We really just need to make sure that the compiler won't reorder the read with the above write.
+                    // RyuJIT won't reorder them today, but more advanced optimizers might.
+                    //
+                    if (Volatile.Read(ref _current) == entry)
```

Comment:
It depends on whether on you are optimizing for x86/x64 or arm/arm64. 

Volatile on x86/x64 does regular load without explicit barrier - it will be faster than the noinline variant. 

Volatile on arm/arm64 has explicit barrier. The it is about measuring whether the explicit barrier is more expensive then a call overhead. I think they will be pretty close, but I do not have a arm/arm64 machine around to check it.